File: src/main/java/twao/assigners/AssignerBuilder.java
package twao.assigners;

import twao.exceptions.UnspecifiedKeyException;
import twao.villages.AllyVillage;
import twao.villages.TargetVillage;
import twao.villages.Village;

import java.util.List;

public class AssignerBuilder {
    /**
     * Determines assigner behavior
     *
     * -. RAM: (concrete attack without noble) Pick ALLY village that is the farthest from [#relativityPoint] and link it with target village that
     * distance between it and given village is the shortest. Repeat.
     *
     * -. REVERSED_RAM: (concrete attack without noble) Pick TARGET village that is the closest to [#relativityPoint and link it with ally village
     * that distance between it and given village is the shortest. Repeat.
     *
     * -. FAKE_RAM: (fake attack without noble) Act like [type.RAM].
     *
     * -. NOBLE: (concrete attack with noble) Pick TARGET village that is the closest to [#relativityPoint] and link it with ally village
     * that distance between it and given village is the shortest. Repeat.
     *
     * -. FAKE_NOBKE: (fake attack with noble) Act like [type.NOBLE].
     */
    public enum AssignerType {
        RAM,
        REVERSED_RAM,
        FAKE_RAM,
        NOBLE,
        FAKE_NOBLE
    }

    private List<TargetVillage> targets;
    private List<AllyVillage> attackingVillages;
    private Village relativityPoint;
    private AssignerType type;
    private int maxNobleRange = -1;

    /**
     * Requires [#targets], [#attackingVillages], [#relativityPoint], [#type] to be set.
     * If assigning (fake)nobles [#maxNobleRange] is also obligatory.
     *
     * @return Concrete Assigner instance
     */
    public Assigner build() throws UnspecifiedKeyException {
        if (targets == null || attackingVillages == null || relativityPoint == null || type == null) {
            throw new UnspecifiedKeyException();
        }
        else if ( (type == AssignerType.NOBLE || type == AssignerType.FAKE_NOBLE) && maxNobleRange == -1 ) {
            throw new UnspecifiedKeyException("Max noble range is not specified.");
        }


        Assigner assigner = null;
        switch (type) {
            case RAM:
                assigner = new Assigner(targets, attackingVillages, relativityPoint, false, new RamAssigning());
                break;
            case REVERSED_RAM:
                assigner = new Assigner(targets, attackingVillages, relativityPoint, false, new ReversedRamAssigning());
                break;
            case FAKE_RAM:
                assigner = new Assigner(targets, attackingVillages, relativityPoint, true, new RamAssigning());
                break;
            case NOBLE:
                assigner = new Assigner(targets, attackingVillages, relativityPoint, false, new NobleAssigning(maxNobleRange));
                break;
        }

        return assigner;
    }

    /** Obligatory */
    public AssignerBuilder setTargets(List<TargetVillage> targets) {
        this.targets = targets;
        return this;
    }

    /** Obligatory */
    public AssignerBuilder setAttackingVillages(List<AllyVillage> attackingVillages) {
        this.attackingVillages = attackingVillages;
        return this;
    }

    /** Obligatory */
    public AssignerBuilder setRelativityPoint(Village relativityPoint) {
        this.relativityPoint = relativityPoint;
        return this;
    }

    /** Obligatory */
    public AssignerBuilder setType(AssignerType type) {
        this.type = type;
        return this;
    }

    /** Used only when assigning (fake)nobles */
    public AssignerBuilder setMaxNobleRange(int maxNobleRange) {
        this.maxNobleRange = maxNobleRange;
        return this;
    }
}

File: src/main/java/twao/assigners/AssignerBuilder.kt
package twao.assigners

import twao.villages.AllyVillage
import twao.villages.TargetVillage
import twao.villages.Village

class AssignerBuilder {
    /**
     * Determines assigner behavior
     *
     * ---------------------------------------------------------------------------------------------------------------
     * [RAM]:
     * output assigment: Concrete attack without noble
     * behavior: Pick ALLY village that is the farthest from referencePoint and link it with nearest target. Repeat.
     *---------------------------------------------------------------------------------------------------------------
     * [REVERSED_RAM]:
     * output assigment: Concrete attack without noble
     * Pick TARGET village that is closest to the referencePoint and link it with nearest ally village. Repeat.
     *---------------------------------------------------------------------------------------------------------------
     * [FAKE_RAM]:
     * output assigment: Fake attack without noble
     * Pick ALLY village that is the farthest from referencePoint and link it with nearest target. Repeat.
     *---------------------------------------------------------------------------------------------------------------
     * [NOBLE]:
     * output assigment: Concrete attack with noble
     * Pick TARGET village that is closest to the relativityPoint and link it with nearest ally village. Repeat.
     *---------------------------------------------------------------------------------------------------------------
     * [FAKE_NOBKE]:
     * output assigment: Fake attack with noble
     * Pick TARGET village that is closest to the relativityPoint and link it with nearest ally village. Repeat.
     * ---------------------------------------------------------------------------------------------------------------
     */
    enum class AssignerType {
        RAM,
        REVERSED_RAM,
        FAKE_RAM,
        NOBLE,
        FAKE_NOBLE
    }

    var targets: MutableList<TargetVillage>? = null
    var resources: MutableList<AllyVillage>? = null
    var mainReferencePoint: Village? = null
    var type: AssignerType? = null
    var maxNobleRange: Int? = null

    /**
     * Requires [targets], [resources], [mainReferencePoint], [type] to be set.
     * !!! If assigning (fake)nobles [maxNobleRange] is also obligatory. !!!
     */
    fun build(): Assigner? {
        if (targets == null || resources == null || mainReferencePoint == null || type == null)
            return null
        if (type!! in listOf(AssignerType.NOBLE, AssignerType.FAKE_NOBLE)
                && maxNobleRange == null)
            return null

        return when (type) {
            AssignerType.RAM -> StandardRamAssigner(
                    targets!!,
                    resources!!,
                    mainReferencePoint!!,
                    isAssigningFakes = false
            )
            AssignerType.REVERSED_RAM -> ReversedRamAssigner(
                    targets!!,
                    resources!!,
                    mainReferencePoint!!,
                    isAssigningFakes = false
            )
            AssignerType.FAKE_RAM -> StandardRamAssigner(
                    targets!!,
                    resources!!,
                    mainReferencePoint!!,
                    isAssigningFakes = true
            )
            AssignerType.NOBLE -> NobleAssigner(
                    targets!!,
                    resources!!,
                    mainReferencePoint!!,
                    isAssigningFakes = false,
                    maxNobleRange = maxNobleRange!!
            )
            AssignerType.FAKE_NOBLE -> NobleAssigner(
                    targets!!,
                    resources!!,
                    mainReferencePoint!!,
                    isAssigningFakes = true,
                    maxNobleRange = maxNobleRange!!
            )
            else -> null
        }
    }

    fun targets(targets: MutableList<TargetVillage>)
        = apply { this.targets = targets }

    fun resources(resources: MutableList<AllyVillage>)
        = apply { this.resources = resources }

    fun mainReferencePoint(referencePoint: Village)
        = apply { this.mainReferencePoint = referencePoint }

    fun type(type: AssignerType)
        = apply { this.type = type }

    fun maxNobleRange(range: Int)
        = apply { this.maxNobleRange = range }
}